pkg/controller/component/vars_test.go

```diff
@@ -2238,5 +2242,326 @@ var _ = Describe("vars", func() {
+					},
+				}
+				reader := &mockReader{
+					cli: testCtx.Cli,
+					objs: []client.Object{
+						&corev1.Secret{
+							ObjectMeta: metav1.ObjectMeta{
+								Namespace: testCtx.DefaultNamespace,
+								Name:      constant.GenerateAccountSecretName(synthesizedComp.ClusterName, synthesizedComp.Name, "credential"),
+							},
+							Data: map[string][]byte{
+								constant.AccountPasswdForSecret: []byte("password"),
+							},
+						},
+					},
+				}
+				templateVars, envVars, err := ResolveTemplateNEnvVars(testCtx.Ctx, reader, synthesizedComp, vars)
+				Expect(err).Should(Succeed())
+				Expect(templateVars).ShouldNot(HaveKey("password"))
+				checkEnvVarWithValueFrom(envVars, "password", &corev1.EnvVarSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: reader.objs[0].GetName(),
+						},
+						Key: constant.AccountPasswdForSecret,
+					},
+				})
+			})
+
+			It("depends on credential-vars", func() {
+				vars := []appsv1alpha1.EnvVar{
+					{
+						Name: "raw",
+						ValueFrom: &appsv1alpha1.VarSource{
+							CredentialVarRef: &appsv1alpha1.CredentialVarSelector{
+								ClusterObjectReference: appsv1alpha1.ClusterObjectReference{
+									Name:     "credential",
+									Optional: required(),
+								},
+								CredentialVars: appsv1alpha1.CredentialVars{
+									Password: &appsv1alpha1.VarRequired,
+								},
+							},
+						},
+						Expression: expp("panic"),
+					},
+					{
+						Name:       "password",
+						Expression: expp("{{ .raw }}"), // depends on $raw which is a credential-var
+					},
+				}
+				reader := &mockReader{
+					cli: testCtx.Cli,
+					objs: []client.Object{
+						&corev1.Secret{
+							ObjectMeta: metav1.ObjectMeta{
+								Namespace: testCtx.DefaultNamespace,
+								Name:      constant.GenerateAccountSecretName(synthesizedComp.ClusterName, synthesizedComp.Name, "credential"),
+							},
+							Data: map[string][]byte{
+								constant.AccountNameForSecret:   []byte("username"),
+								constant.AccountPasswdForSecret: []byte("password"),
+							},
+						},
+					},
+				}
+				_, _, err := ResolveTemplateNEnvVars(testCtx.Ctx, reader, synthesizedComp, vars)
+				Expect(err).ShouldNot(Succeed())
+				Expect(err.Error()).Should(And(ContainSubstring("map has no entry for key"), ContainSubstring("raw")))
+			})
+
+			It("depends on intermediate values", func() {
+				vars := []appsv1alpha1.EnvVar{
+					{
+						Name:       "endpoint",
+						Expression: expp("{{ .host }}:{{ .port }}"),
+					},
+					{
+						Name:       "host",
+						Value:      "localhost",
+						Expression: expp("127.0.0.1"),
+					},
+					{
+						Name:  "port",
+						Value: "12345",
+					},
+				}
+				templateVars, envVars, err := ResolveTemplateNEnvVars(testCtx.Ctx, nil, synthesizedComp, vars)
+				Expect(err).Should(Succeed())
+				Expect(templateVars).Should(HaveKeyWithValue("endpoint", "localhost:12345"))
```

Comment:
seems to be an unexpected result, how about resolved multiple-passes?

Reply:
Consider such an example:
```yaml
vars:
  - name: a
    value: 1
    expression: {{.a}}+{{.b}}
  - name: b
    value: 1
    expression: {{.a}}+{{.b}}
```
Even if we employ multiple passes to evaluate, it will never have a stable result.

Reply:
We can exclue the possibility of circular dependencies. If a circular dependency is detected, an error should be thrown.

Reply:
Since we do not parse and understand the content of the expressions (it's too complicated to do that), we have no idea about whether there are any circular dependencies, we can only check whether further iteration is needed by comparing the results of the two previous evaluations. As for whether the iteration can be ended or how many more times it‘s needed, there is no way for us to know it in advance. It may keep iterating indefinitely, unless we set an upper limit on the number of iterations.

Reply:
Fine, i will submit another PR to improve it.